### ISSUE TITLE
[TW-73] new special-tag +lock

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#132]
_David Patrick on 2013-04-21T04:39:40Z says:_ 

 (was; new special-tags +norm and +nodo)

There may be some tasks that you want to keep around for a long while, either as big meta-tasks or as templates or as utility tasks. For whatever reason, none of which you want to see deleted or marked done prematurely.

Using the +norm and/or +nodo makes it impossible for taskwarrior to remove (delete) or mark as done (do), period. 
In order to delete a +norm (no-remove) task or mark a +nodo (no-done) task as done, you have to explicitly remove the special-tag first. Trying to do so should result in an error that says something like; "this task has the +norm special-tag. to delete it you  must remove the tag first". This ensures that such tasks won't be accidentally deleted or marked done.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2013-04-21T04:39:40Z
Modified: 2015-03-14T16:02:12Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2013-04-21T04:45:21Z says:_ 

 I have this, or something similar in my notes as @+lock@ which prevents any modification.  Who is Norm?

**Comment by taskwarrior** [comment]
_David Patrick on 2013-04-21T14:42:23Z says:_ 

 Paul Beckingham wrote:
> Who is Norm?

oh.. he's the idiot who came up with this idea, wanted "screen credit"! 

+lock is even better!

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2013-04-21T14:58:33Z says:_ 

 Maybe Norm would like to select a target version when this should appear.

**Comment by taskwarrior** [comment]
_Thomas Sullivan on 2013-05-08T12:47:26Z says:_ 

 +lock is a nice tag idea - I tend to set a due date of the end of quarter and then review all of those long term tasks as part of a cleanup/organization schedule.  

Is there already a config item for "do not delete without prompt=true/false" type thing?  Then again, delete is not equal to done.  When I have one I want around I tend to make it recurring so when I finish one, I get another instantly and then have a record of how many times I've done it, per se. ( One of my stupid weekly tasks involves submitting a timesheet of what I've pretended to work on all week long... :P )

**Comment by taskwarrior** [comment]
_Tullio Facchinetti on 2013-05-08T20:51:58Z says:_ 

 Paul Beckingham wrote:
> I have this, or something similar in my notes as @+lock@ which prevents any modification.  Who is Norm?

To familiarize with tw internals, I added this feature with a small patch.

It affects the following commands: annotate, append, delete, denotate, done, edit, modify, prepend, start, stop.

The behavior is simple: for each task that is tagged with a +lock tag, a message like

  Task {ID} locked: skipping.

and the command is not applied to that task with the mentioned ID.

I think it should address the issue, at least partially.

I say partially, since I have no idea how the +locked tag should affect the behavior of following commands: pull, merge, sync.

Any hint is welcome.

PS: bring my greetings to Norm.

**Comment by taskwarrior** [comment]
_Tullio Facchinetti on 2013-05-08T20:57:11Z says:_ 

 As a reminder: it would be nice to be able to specify the set of commands affected by the +lock tag in the configuration file.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2013-05-09T03:56:42Z says:_ 

 Tullio Facchinetti wrote:
> As a reminder: it would be nice to be able to specify the set of commands affected by the +lock tag in the configuration file.

I think we can do that already.  There is the notion of "read-only" and "write" commands.  I think @+lock@ should inhibit all write commands, so rather than maintaining a list, we simply apply the whole category.

Norm says hi.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2014-12-27T15:33:08Z says:_ 

 Tulio, your patch is missing.  We must have lost it when converting to the new bug system.  Do you have it?  If you have it available, I will get it into 2.4.1, otherwise I will recreate it.

I'm wondering if instead of +lock, it should be +readonly.  Either way is good.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2015-03-14T16:02:12Z says:_ 

 Closing this.  In the interest of keeping the Taskwarrior core lean, there are improvements that can and should be achieved using hook scripts.  This is one.
